dubbo-remoting/dubbo-remoting-zookeeper/src/main/java/org/apache/dubbo/remoting/zookeeper/support/AbstractZookeeperTransporter.java

```diff
@@ -0,0 +1,130 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.dubbo.remoting.zookeeper.support;
+
+import org.apache.dubbo.common.Constants;
+import org.apache.dubbo.common.URL;
+import org.apache.dubbo.common.logger.Logger;
+import org.apache.dubbo.common.logger.LoggerFactory;
+import org.apache.dubbo.remoting.zookeeper.ZookeeperClient;
+import org.apache.dubbo.remoting.zookeeper.ZookeeperTransporter;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * 2019/1/10
+ */
+public abstract class AbstractZookeeperTransporter implements ZookeeperTransporter {
+    static final Logger logger = LoggerFactory.getLogger(ZookeeperTransporter.class);
+    Map<String, ZookeeperClientData> zookeeperClientMap = new HashMap<String, ZookeeperClientData>();
+
+    /**
+     * share connnect for registry, metadata, etc..
+     *
+     * @param url
+     * @return
+     */
+    public ZookeeperClient connect(URL url) {
+        ZookeeperClientData clientData;
+        List<String> addressList = getURLBackupAddress(url);
+        // The field define the zookeeper server , including protocol, host, port, username, password
+        if ((clientData = fetchAndUpdateZookeeperClientCache(url, addressList)) != null) {
+            logger.info("ZookeeperTransporter.connnect result from map:" + clientData);
+            return clientData.getZookeeperClient();
+        }
+        ZookeeperClient zookeeperClient = null;
+        synchronized (zookeeperClientMap) {
```

Comment:
Above calls uses method **fetchAndUpdateZookeeperClientCache** to get zookeeper client which internally use **zookeeperClientMap ** and here we are trying to get lock on zookeeperClientMap, is there any reason in one place to have lock and another avoid lock? Is there I am missing anything ?

Reply:
It is still need this lock, bacause I want make the ZookeeperClient creation is single at any time.